keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+- As an application developer, I don't want to study the nitty-gritty details of different runtime
+  implementations.
+- As an application developer, I want my application to be portable across clusters that use similar
+  but different variants of a "class" of runtimes.
+- As an application operator, I want to know precisely how the application is running, and which
+  features are being used.
+- As an application operator, I want pods to fail fast when trying to use an optional feature that
+  isn't supported by the chosen runtime.
+- As a security operator, I want to ensure that pods are running with the security features they
+  claim.
+
+## Proposal
+
+The initial design includes:
+
+- `RuntimeClass` API resource definition
+- `RuntimeClass` pod field for specifying the RuntimeClass the pod should be run with
+- Kubelet implementation for fetching & interpreting the RuntimeClass
+- CRI API & implementation for passing along the RuntimeClass parameters
+
+A subsequent update will include the API for describing supported features and an admission
+controller for performing validation & defaulting.
+
+### API
+
+`RuntimeClass` is a new cluster-scoped resource in the `node.k8s.io` API group.
+
+> _The `node.k8s.io` API group would eventually hold the Node resource when `core` is retired.
+> Alternatives considered: `runtime.k8s.io`, `cluster.k8s.io`_
+
+There are 3 categories of fields that may be included in the `RuntimeClassSpec`:
+
+1. **Runtime setup** - These fields control the behavior of the runtime in some way,
+   including configuring arbitrary options in the runtime (`Parameters`), and eventually specifying
+   resource overhead. This is the focus of this initial proposal.
+2. **Compatibility validation** - These fields declare support for various features, and pods are
+   validated against them in admission. This category, along with the admission controller, will be
+   addressed in a subsequent update to this proposal.
+3. **Informational** - These fields surface semantic properties of the runtime to guide runtime
+   selection. Examples could include hardware virtualization, stability, and I/O performance
+   class. These fields are out-of-scope for the initial proposal.
+
+#### API Specification
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+type RuntimeClass struct {
+    metav1.TypeMeta
+    // ObjectMeta minimally includes the RuntimeClass name, which is used to reference the class.
+    // Namespace should be left blank.
+    metav1.ObjectMeta
+
+    Spec RuntimeClassSpec
+}
+
+type RuntimeClassSpec struct {
+    // Parameters holds opaque parameters that are interpreted by the runtime (CRI implementation).
+    Parameters map[string]string
+}
+```
+
+The runtime is selected by the pod by specifying the RuntimeClass in the PodSpec. Once the pod is
+scheduled, the RuntimeClass cannot be changed.
```

Comment:
should be enforced in the admission.

Reply:
I think this would be enforced in validation (https://github.com/kubernetes/kubernetes/blob/44073e6f43af29c0fb8cad66593d850435c6e2be/pkg/apis/core/validation/validation.go#L3434)

Reply:
Yes, restrictions on pod update will need to be enforced in validation (as it won't go through admission)